src/lem/coroutine/toplevel.rs

```diff
@@ -0,0 +1,494 @@
+        let dummy = [
+            AllocatedPtr::alloc_infallible::<_, _, Tag>(&mut cs, || unreachable!()),
+            AllocatedPtr::alloc_infallible::<_, _, Tag>(&mut cs, || unreachable!()),
+            AllocatedPtr::alloc_infallible::<_, _, Tag>(&mut cs, || unreachable!()),
+            AllocatedPtr::alloc_infallible::<_, _, Tag>(&mut cs, || unreachable!()),
+            AllocatedPtr::alloc_infallible::<_, _, Tag>(&mut cs, || unreachable!()),
+            AllocatedPtr::alloc_infallible::<_, _, Tag>(&mut cs, || unreachable!()),
+        ];
+        factorial_circuit
+            .supernova_synthesize(&mut cs, &dummy)
+            .unwrap();
+        expect!("1772").assert_eq(&cs.num_constraints().to_string());
+
+        let index = toplevel.0.get_index_of(&even).unwrap();
+        let even_circuit = CoroutineCircuit::<_, _, ToplevelQuery<_>>::blank(
+            index,
+            s.clone(),
+            1,
+            toplevel.clone(),
+        );
+        let mut cs = BenchCS::new();
+        let dummy = [
+            AllocatedPtr::alloc_infallible::<_, _, Tag>(&mut cs, || unreachable!()),
+            AllocatedPtr::alloc_infallible::<_, _, Tag>(&mut cs, || unreachable!()),
+            AllocatedPtr::alloc_infallible::<_, _, Tag>(&mut cs, || unreachable!()),
+            AllocatedPtr::alloc_infallible::<_, _, Tag>(&mut cs, || unreachable!()),
+            AllocatedPtr::alloc_infallible::<_, _, Tag>(&mut cs, || unreachable!()),
+            AllocatedPtr::alloc_infallible::<_, _, Tag>(&mut cs, || unreachable!()),
+        ];
+        even_circuit.supernova_synthesize(&mut cs, &dummy).unwrap();
+        expect!("1772").assert_eq(&cs.num_constraints().to_string());
+    }
+
+    #[test]
+    fn prove_sum_list_coroutine() {
+        let build_list = func!(build_list(n): 1 => {
+            let zero = Num(0);
+            let n_is_zero = eq_val(n, zero);
+            if n_is_zero {
+                let nil: Expr::Nil;
+                return (nil)
+            }
+            let one = Num(1);
+            let m = sub(n, one);
+            let ys = QUERY(build_list, m);
+            let xs: Expr::Cons = cons2(n, ys);
+            return (xs)
+        });
+        let sum_list = func!(sum_list(xs): 1 => {
+            match xs.tag {
+                Expr::Nil => {
+                    let zero = Num(0);
+                    return (zero)
+                }
+                Expr::Cons => {
+                    let (n, ys) = decons2(xs);
+                    let m = QUERY(sum_list, ys);
+                    let res = add(n, m);
+                    return (res)
+                }
+            }
+        });
+        let main = func!(main(n): 1 => {
+            let xs = QUERY(build_list, n);
+            let n = QUERY(sum_list, xs);
+            return (n)
+        });
+        let main_sym = user_sym("main");
+        let build_list_sym = user_sym("build_list");
+        let sum_list_sym = user_sym("sum_list");
```

Comment:
We should use the Lurk naming convention for symbols, in order to promote consistent style, so `build-list` and `sum-list`.